**Рассолова Маргарита Сергеевна, НММбд-02-22**

In [3]:
import pandas as pd

1. Откройте файл с таблицей. Выведите последние 10 строк.

In [4]:
file= 'chipotle.tsv'

In [6]:
df = pd.read_csv(file, sep='\t') # Загрузка данных из файла в DataFrame

In [7]:
last_10_orders = df.tail(10)

In [8]:
print(last_10_orders)

      order_id  quantity            item_name  \
4612      1831         1        Carnitas Bowl   
4613      1831         1                Chips   
4614      1831         1        Bottled Water   
4615      1832         1   Chicken Soft Tacos   
4616      1832         1  Chips and Guacamole   
4617      1833         1        Steak Burrito   
4618      1833         1        Steak Burrito   
4619      1834         1   Chicken Salad Bowl   
4620      1834         1   Chicken Salad Bowl   
4621      1834         1   Chicken Salad Bowl   

                                     choice_description item_price  
4612  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...     $9.25   
4613                                                NaN     $2.15   
4614                                                NaN     $1.50   
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75   
4616                                                NaN     $4.45   
4617  [Fresh Tomato Salsa, [Rice, Black Beans,

Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк, какие столбцы?

- Данные состоят из id продукта, его количества, названия, состава и цены.

2. Ответьте на вопросы:
Сколько заказов попало в выборку?
Сколько уникальных категорий товара было куплено? (item_name)

In [9]:
total_orders_last_10 = last_10_orders['order_id'].nunique()


In [10]:
unique_categories_last_10 = last_10_orders['item_name'].nunique()

In [11]:
print(f"Количество заказов в последних 10 строках: {total_orders_last_10}")
print(f"Количество уникальных категорий товара в последних 10 строках: {unique_categories_last_10}")

Количество заказов в последних 10 строках: 4
Количество уникальных категорий товара в последних 10 строках: 7


3.  Есть ли в данных пропуски? В каких колонках?

In [12]:
columns_with_nulls = df.columns[df.isnull().any()]

In [13]:
print("Колонки с пропусками:", columns_with_nulls)

Колонки с пропусками: Index(['choice_description'], dtype='object')


Заполните пропуски пустой строкой для строковых колонок и нулём для числовых.

In [14]:
df.fillna({'column_name_1': '', 'column_name_2': 0}, inplace=True)

4. Посмотрите внимательнее на колонку с ценой товара. Какого она типа? Создайте новую колонку так, чтобы в ней цена была числом.

Для этого попробуйте применить функцию-преобразование к каждой строке вашей таблицы (для этого есть соответствующая функция).

In [15]:
df['item_price'] = pd.to_numeric(df['item_price'].str.replace('$', ''), errors='coerce')

<ipython-input-15-f192e3e496a9>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df['item_price'] = pd.to_numeric(df['item_price'].str.replace('$', ''), errors='coerce')


Какая средняя/минимальная/максимальная цена у товара?

In [16]:
# Средняя цена товара
average_price = df['item_price'].mean()

In [17]:
# Минимальная цена товара
min_price = df['item_price'].min()

In [18]:
# Максимальная цена товара
max_price = df['item_price'].max()

In [19]:
print(f"Средняя цена: {average_price:.2f}")
print(f"Минимальная цена: {min_price:.2f}")
print(f"Максимальная цена: {max_price:.2f}")

Средняя цена: 7.46
Минимальная цена: 1.09
Максимальная цена: 44.25


Удалите старую колонку с ценой.

In [20]:
df = df.drop(columns=['item_price'])


5. Какие 5 товаров были самыми дешёвыми и самыми дорогими? (по item_name)

Для этого будет удобно избавиться от дубликатов и отсортировать товары. Не забудьте про количество товара.

In [24]:
# Загрузка данных из файла в DataFrame
df = pd.read_csv(file_path, sep='\t')


In [25]:
# Преобразование цены товара в числовой формат
df['item_price'] = df['item_price'].apply(lambda x: float(x[1:]))

# Создание новой колонки 'total_price' для учета количества товара
df['total_price'] = df['quantity'] * df['item_price']

In [26]:
# Группировка данных по 'item_name', нахождение минимальной и максимальной цены
cheapest_items = df.groupby('item_name')['total_price'].min().reset_index()
expensive_items = df.groupby('item_name')['total_price'].max().reset_index()

In [27]:
# Сортировка по цене
cheapest_items = cheapest_items.sort_values(by='total_price').head(5)
expensive_items = expensive_items.sort_values(by='total_price', ascending=False).head(5)

In [28]:
print("Самые дешёвые товары:")
print(cheapest_items)

print("\nСамые дорогие товары:")
print(expensive_items)

Самые дешёвые товары:
            item_name  total_price
6       Bottled Water         1.09
9         Canned Soda         1.09
10  Canned Soft Drink         1.25
37      Side of Chips         1.69
23              Chips         1.99

Самые дорогие товары:
                       item_name  total_price
24  Chips and Fresh Tomato Salsa       663.75
6                  Bottled Water       150.00
18               Chicken Burrito       140.00
37                 Side of Chips       108.16
11                 Carnitas Bowl       105.75


6. Сколько раз клиенты покупали больше 1 Chicken Bowl (item_name)?

In [29]:
chicken_bowl_orders = df[df['item_name'] == 'Chicken Bowl']
more_than_1_chicken_bowl_orders = chicken_bowl_orders[chicken_bowl_orders['quantity'] > 1]['order_id'].nunique()


In [31]:
print(f"Количество раз, когда клиенты покупали больше 1 Chicken Bowl: {more_than_1_chicken_bowl_orders}")

Количество раз, когда клиенты покупали больше 1 Chicken Bowl: 33


7. Какой средний чек у заказа? Сколько в среднем товаров покупают?

Если необходимо провести вычисления в терминах заказов, то будет удобно сгруппировать строки по заказам и посчитать необходимые статистики.

In [32]:
# Группировка данных по заказам и подсчет среднего чека и среднего количества товаров
average_order_stats = df.groupby('order_id').agg({
    'item_price': 'sum',
    'quantity': 'sum'
}).mean()

In [36]:
average_order_price = average_order_stats['item_price']
average_items_per_order = int(average_order_stats['quantity'])

In [37]:
print(f"Средний чек у заказа: ${average_order_price:.2f}")
print(f"Среднее количество товаров: {average_items_per_order:.2f}")

Средний чек у заказа: $18.81
Среднее количество товаров: 2.00


8. Сколько заказов содержали ровно 1 товар?

In [38]:
# Подсчет уникальных заказов и их количества
orders_count = df.groupby('order_id')['quantity'].count()


In [39]:
# Подсчет заказов, содержащих ровно 1 товар
orders_with_one_item = orders_count[orders_count == 1].count()

In [40]:
print(f"Количество заказов, содержащих ровно 1 товар: {orders_with_one_item}")

Количество заказов, содержащих ровно 1 товар: 128


9. Какая самая популярная категория товара?

In [41]:
# Подсчет количества заказов для каждой категории товара
popular_categories = df.groupby('item_name')['order_id'].count()

In [42]:
# Нахождение самой популярной категории товара
most_popular_category = popular_categories.idxmax()

In [43]:
print(f"Самая популярная категория товара: {most_popular_category}")

Самая популярная категория товара: Chicken Bowl


10. Какие виды Burrito существуют? Какой из них чаще всего покупают? Какой из них самый дорогой?

In [44]:
# Фильтрация данных по категории 'Burrito'
burrito_data = df[df['item_name'].str.contains('Burrito')]

In [45]:
# Определение уникальных видов Burrito
burrito_types = burrito_data['item_name'].unique()

# Вывод видов Burrito
print("Виды Burrito:", burrito_types)


Виды Burrito: ['Steak Burrito' 'Chicken Burrito' 'Barbacoa Burrito' 'Carnitas Burrito'
 'Veggie Burrito' 'Burrito']


In [46]:
# Подсчет частоты покупок для каждого вида Burrito
burrito_frequency = burrito_data.groupby('item_name')['order_id'].count()

# Нахождение самого часто покупаемого вида Burrito
most_ordered_burrito = burrito_frequency.idxmax()

In [48]:
print(f"Самый часто покупаемый вид Burrito: {most_ordered_burrito}")


Самый часто покупаемый вид Burrito: Chicken Burrito


In [49]:
# Нахождение самого дорогого вида Burrito
most_expensive_burrito = burrito_data.loc[burrito_data['item_price'].idxmax()]['item_name']

In [50]:
print(f"Самый дорогой вид Burrito: {most_expensive_burrito}")

Самый дорогой вид Burrito: Chicken Burrito


11. В каком количестве заказов есть товар, который стоит более 40% от суммы всего чека?

In [51]:
# Группировка данных по номеру заказа и подсчет общей стоимости каждого заказа
order_totals = df.groupby('order_id')['item_price'].sum()

In [52]:
# Фильтрация заказов, в которых есть товары, стоимость которых превышает 40% от суммы заказа
filtered_orders = order_totals[order_totals > 0.4 * order_totals]

In [53]:
# Вывод количества таких заказов
print(f"Количество заказов, в которых есть товар, стоимость которого превышает 40% от суммы заказа: {len(filtered_orders)}")

Количество заказов, в которых есть товар, стоимость которого превышает 40% от суммы заказа: 1834


12. Предположим, что в данных была ошибка и Diet Coke (choice_description), который стоил $1.25, должен был стоить 1.35. Скорректируйте данные в таблицы и посчитайте, на какой процент больше денег было заработано с этого товара. Не забывайте, что количество товара не всегда равно 1.

In [54]:
# Фильтрация данных для товара "Diet Coke"
diet_coke_data = df[df['item_name'] == 'Diet Coke']

In [55]:
# Расчет текущей выручки и количества товара
current_revenue = (diet_coke_data['quantity'] * diet_coke_data['item_price']).sum()
current_quantity = diet_coke_data['quantity'].sum()

In [59]:
# Проверка на ноль текущей выручки
if current_revenue == 0:
    print("Текущая выручка равна нулю.")
else:
    # Исправление цены для товара "Diet Coke"
    corrected_price = 1.35
    df.loc[df['item_name'] == 'Diet Coke', 'item_price'] = corrected_price

    # Расчет новой выручки после коррекции
    updated_revenue = (diet_coke_data['quantity'] * corrected_price).sum()

    # Расчет процента изменения выручки
    percentage_change = ((updated_revenue - current_revenue) / current_revenue) * 100

Текущая выручка равна нулю.


In [60]:
print(f"Текущая выручка: ${current_revenue:.2f}")
print(f"Новая выручка после изменения цены: ${updated_revenue:.2f}")
print(f"Процент изменения выручки: {percentage_change:.2f}%")

Текущая выручка: $0.00
Новая выручка после изменения цены: $0.00
Процент изменения выручки: nan%
